cluster/rbac/manager.go

```diff
@@ -25,14 +29,41 @@ import (
 var ErrBadRequest = errors.New("bad request")
 
 type Manager struct {
-	authZ  authorization.Controller
+	authZ  *rbac.RBAC
 	logger logrus.FieldLogger
 }
 
-func NewManager(authZ authorization.Controller, logger logrus.FieldLogger) *Manager {
+func NewManager(authZ *rbac.RBAC, logger logrus.FieldLogger) *Manager {
 	return &Manager{authZ: authZ, logger: logger}
 }
 
+func (m *Manager) PreApplyFilter(req *cmd.ApplyRequest) error {
+	if req.Type != cmd.ApplyRequest_TYPE_UPSERT_ROLES_PERMISSIONS {
+		return nil
+	}
+
+	subreq := &cmd.CreateRolesRequest{}
+	if err := json.Unmarshal(req.SubCommand, req); err != nil {
+		return fmt.Errorf("%w: %w", ErrBadRequest, err)
+	}
+
+	// don't allow to create roles if there is already a role present
+	if subreq.RoleCreation {
+		names := make([]string, 0, len(subreq.Roles))
+		for name := range subreq.Roles {
+			names = append(names, name)
+		}
+		roles, err := m.authZ.GetRoles(names...)
+		if err != nil {
+			return err
+		}
+		if len(roles) > 0 {
+			return backoff.Permanent(fmt.Errorf("%w: roles already exist", ErrBadRequest))
```

Comment:
currently the request is repeated a couple of times. When wrapping this error in backoff.permanent this is fixed

Reply:
we shall not use backoff here, there is no retries here , the backoff.permanent shall be used in above layer on `Apply()` 

does it repeat as create or update ?